### ISSUE TITLE
Stroking simple spline (three points) results in strange sharp edges at endpoints

### ISSUE THREAD

**Issue by ctrlcctrlv** [fontforge/fontforge#4073]
![2019-12-20-170011_2786x1517_scrot](https://user-images.githubusercontent.com/838783/71242932-0bc17000-234a-11ea-99f4-7a30ccff2bfc.png)
![2019-12-20-170106_2786x1517_scrot](https://user-images.githubusercontent.com/838783/71242989-2c89c580-234a-11ea-9b82-6c7c1c6db53b.png)

In the grand scheme of things, this problem is small;
I'm hesitant to even call it a bug at all.
But the stroker is new, and the testers are few,
So report it I must, for to do so is just.

# Required settings Ⓐ

![2019-12-20-170559_1088x1147_scrot](https://user-images.githubusercontent.com/838783/71243234-dff2ba00-234a-11ea-8811-8b0b453d1297.png)

Note: Remove Overlap setting doesn't matter.

# Required settings Ⓑ

Is this two issues in one?
The contour is the same,
But the results more seriously wrong.
I feel they're related.

![2019-12-20-171017_1088x1147_scrot](https://user-images.githubusercontent.com/838783/71243513-76bf7680-234b-11ea-8905-eb91a1124acf.png)
![2019-12-20-171041_1768x1517_scrot](https://user-images.githubusercontent.com/838783/71243534-8343cf00-234b-11ea-8e0c-34d947c6fdff.png)

# File

[uni0315.sfd](https://github.com/fontforge/fontforge/files/3987555/uni0315.zip)

**Comment by skef** [comment]
The second, calligraphic output contour is unusual looking but not an error. I added bevel caps as something that would be the simplest option in many cases. This particular result is due to the close alignment of the curve with the nib edge, such that the "drawing corner" changes at the start of the source curve. 

So basically: bevel caps and nibs with flat edges will sometimes yield unattractive/non-simple results, so a butt cap is likely a better choice if a) your nib has edges and b) you care. 

I'm less content explaining away the first case, but "explain away" might end up being the reality of it. Recall #4007, in which we discussed the issue of "visible cusps". That is, somewhat counter-intuitively, the cause of this sub-optimal contour. 

It's counter-intuitive because the curve of the source spline is starting out so gently. Unfortunately that gentleness is counter-acted by the even more gentle curvature of the flattest portion of the very eccentric (in the ellipse sense) nib, which because of the angle match is what is drawing the curve to start with. You don't see the effect with a "nib" cap because a nib (always) covers its own cusps, but the round cap is not large enough to do so completely. 

The result is therefore analogous to:

![68118798-baeada80-feb5-11e9-80b5-a8cb28c926ae](https://user-images.githubusercontent.com/6175836/71248750-7e8e0500-22d0-11ea-95a7-1cce8de20810.png)

(And indeed, if you choose a "butt" rather than a "round" cap the issue is more pronounced.) 

Given that I have no systematic fix for visible cusps, I'm not sure what I can constructively do about this other than warn about the low curvature on the long "sides" of very eccentric elliptical nibs.

**Comment by skef** [comment]
If nothing else I can add a note to the section on bevel caps (which already sort of warns users away from them) about nibs with lines, and document the issue about visible cusps with a couple examples. I should have done the latter before but got documentation fatigue.

**Comment by ctrlcctrlv** [comment]
In practice, the first case is easy to fix with a &ldquo;Simplify&rdquo;, and that's what I did in my font.

I recommend that as a possible workaround to recommend in documentation. You can get attractive results by selecting only the points around the problem area. That's what I did in my latest Chomsky release.

**Comment by skef** [comment]
Ah, OK, so with this particular combination you have to deactivate Simplify to see the problem, and therefore it would only crop up with the default (non-nib) settings if everything were much bigger. 

Since we don't recommend turning Simplify *off* I may say that you can fix some issues with a larger accuracy target or by applying simplify more selectively in post-processing.

**Comment by ctrlcctrlv** [comment]
I discovered this problem by realizing that the curve simply looked _wrong_ with simplify on, like too square, too sharp of a turn. So I turned it off and _voil&agrave;_.

In my font I stroked with simplify off, then selected only the curves in the area and applied Simplify with a larger em value.

**Comment by ctrlcctrlv** [comment]
(By the way, this has nothing at all to do with the issue at hand, but I hope we're good. I got a little too passionate about #4062. We need each other, in a totally platonic non-creepy way.)

**Comment by skef** [comment]
We're good. The scripted-UI problem is particularly complex and fraught. 

And the cusp issue aside it's nice to see the combination of an elliptical nib and an (actually) round cap in use, because that's only available with the new code. Are you enhancing Chomsky or something?

**Comment by ctrlcctrlv** [comment]
Chomsky was [featured in _BoingBoing_](https://boingboing.net/2019/12/19/chomsky-a-blackletter-typefac.html)...so I decided to fix an outstanding issue.

There's a lot more work I could do to Chomsky; @eliheuer is interested in making it variable. [I've toyed with that too.](https://twitter.com/HW_BEAT_THAT/status/1173198769359224833)

Chomsky is, for the most part, made the old fashioned way...the source is available. Spiro is in heavy use like in all my fonts.

Obviously the new stroker is way better than what we had before, so going forward I can use it in more cases.

I have a lot of interest in making an open source version of [Lydian](https://www.myfonts.com/fonts/bitstream/lydian/); I'm especially very keen to draw a Hebrew Lydian. The new stroker makes this possible, though some thought needs to go into it on my end.
